Libraries/ActionSheetIOS/ActionSheetIOS.js

```diff
@@ -49,32 +66,54 @@ var ActionSheetIOS = {
    * - `message` (string) - a message to share
    * - `url` (string) - a URL to share
    *
+   * Returns a Promise that is resolved with an object:
+   *
+   * - `completed` (boolean) - true if share action was completed
+   * - `activityType` (?string) - type of share activity user selected
+   *
+   * NOTE: `failureCallback` and `successCallback` are now deprecated in favour
+   * of promise
+   *
    * NOTE: if `url` points to a local file, or is a base64-encoded
    * uri, the file it points to will be loaded and shared directly.
    * In this way, you can share images, videos, PDF files, etc.
    */
   showShareActionSheetWithOptions(
     options: Object,
-    failureCallback: Function,
-    successCallback: Function
-  ) {
+    failureCallback?: Function,
+    successCallback?: Function
+  ): Promise<ShareSheetResponse> {
     invariant(
       typeof options === 'object' && options !== null,
       'Options must be a valid object'
     );
-    invariant(
-      typeof failureCallback === 'function',
-      'Must provide a valid failureCallback'
-    );
-    invariant(
-      typeof successCallback === 'function',
-      'Must provide a valid successCallback'
-    );
-    RCTActionSheetManager.showShareActionSheetWithOptions(
-      {...options, tintColor: processColor(options.tintColor)},
-      failureCallback,
-      successCallback
-    );
+
+    const promise = RCTActionSheetManager.showShareActionSheetWithOptions({
+      ...options,
+      tintColor: processColor(options.tintColor)
+    });
+
+    if (typeof failureCallback === 'function' || typeof successCallback === 'function') {
+      invariant(
+        typeof failureCallback === 'function',
+        'Must provide a valid failureCallback'
+      );
+      invariant(
+        typeof successCallback === 'function',
+        'Must provide a valid successCallback'
+      );
+      console.warn(
+        'ActionSheetIOS.showShareActionSheetWithOptions returns a Promise ' +
+        'and successCallback function and failureCallback function have been deprecated. ' +
+        'Use showShareActionSheetWithOptions(opts).then().catch() instead.'
+      );
+      promise
+        // $FlowFixMe: Ignores invariants and reports possible undefined value
+        .then((res: ShareSheetResponse) => successCallback(res.completed, res.activityType))
```

Comment:
The only change except the interface in this PR is here. Previously the successCallback was receiving (completed, type) as arguments. Obviously impossible with Promises and an array isn't a good idea as well. Decided to wrap it in an object which also allows us to attach flow.